ssg/build_cpe.py

```diff
@@ -9,55 +9,61 @@
 
 from .constants import oval_namespace
 from .constants import PREFIX_TO_NS
-from .utils import merge_dicts, required_key
+from .utils import required_key
 from .xml import ElementTree as ET
-from .yaml import open_and_macro_expand
 from .boolean_expression import Algebra, Symbol, Function
+from .data_structures import XCCDFEntity
+
 
 class CPEDoesNotExist(Exception):
     pass
 
+
 class ProductCPEs(object):
     """
     Reads from the disk all the yaml CPEs related to a product
     and provides them in a structured way.
     """
 
-    def __init__(self, product_yaml):
-        self.product_yaml = product_yaml
+    def __init__(self):
 
         self.cpes_by_id = {}
         self.cpes_by_name = {}
         self.product_cpes = {}
         self.platforms = {}
         self.algebra = Algebra(symbol_cls=CPEALFactRef, function_cls=CPEALLogicalTest)
 
-        self.load_product_cpes()
-        self.load_content_cpes()
-
     def _load_cpes_list(self, map_, cpes_list):
         for cpe in cpes_list:
             for cpe_id in cpe.keys():
-                map_[cpe_id] = CPEItem(cpe[cpe_id])
+                map_[cpe_id] = CPEItem.get_instance_from_full_dict(cpe[cpe_id])
 
-    def load_product_cpes(self):
+    def load_product_cpes(self, env_yaml):
         try:
-            product_cpes_list = self.product_yaml["cpes"]
-            self._load_cpes_list(self.product_cpes, product_cpes_list)
-
+            product_cpes_list = env_yaml["cpes"]
+            for cpe in product_cpes_list:
+                for cpe_id in cpe.keys():
+                    # these product CPEs defined in product.yml are defined
+                    # differently than CPEs in shared/applicability/*.yml
+                    # therefore we have to place the ID at the place where it is expected
+                    cpe[cpe_id]["id_"] = cpe_id
+                    cpe_item = CPEItem.get_instance_from_full_dict(cpe[cpe_id])
+                    cpe_item.is_product_cpe = True
+                    self.add_cpe_item(cpe_item)
         except KeyError:
-            print("Product %s does not define 'cpes'" % (self.product_yaml["product"]))
+            print("Product %s does not define 'cpes'" % (env_yaml["product"]))
```

Comment:
Why it doesn't raise an exception of some class with a message in the exception?

Reply:
Thanks, fixed.